### PR TITLE
Migrate to Built in HTTP to replace Express

### PR DESCRIPTION
**Pull request by confused-Techie**
This PR does just like the title says, and removes `Express` from the `download` microservice endpoint.

All initial handling of the web request is done within Node's Built in `HTTP` module.

This PR was very largely inspired by #49 as with this merged that would mean the `download` microservice is a dependency free solution.

While the PR linked does provide additional Speed and Size benefits, this PR comes with it's own as well.

Before:
  - Size: 4.03MB
  - Speed: 641ms - 515ms - 521ms

After:
  - Size: 2.13MB
  - Speed: 517ms - 532ms - 602ms
  
So while the performance in size means this microservice is nearly half the size as it was previously, unfortunately the speed advantage isn't that great. Although with that said, the microservice was still faster on average than when using Express.

### CODE REVIEW

**Review by DeeDeeG — COMMENTED** [review]
I can vouch for the general concept of the built-in 'http'/'https' modules of NodeJS working quite well, at least as much as I needed them for simple things I've tried in the past. Such as PR https://github.com/pulsar-edit/package-frontend/pull/49 and [the script](https://github.com/pulsar-edit/ppm/pull/42/files#diff-68b179ff702d03d77bf161e02ca77d1fc14890445b403612de333855aecb670f) I did for that (maybe over-complicated) ppm PR. 👍 for premise.

Questions I have:

- Are these connections meant to be plain HTTP, or is the content privacy of HTTPS relevant here?
  - (It might not be relevant, I suppose, it just discourages snooping and interception/altering the data, but this is not mega sensitive data to begin with? The interception is hypothetically a worry, I guess, due to the ability to replace the links or whatever. But HTTPS is not fool-proof either. Hopefully no-one is intercepting our stuff, but the performance hit of HTTPS is not too bad if any either, at least it wasn't when I was doing https://github.com/pulsar-edit/package-frontend/pull/49.)
  - Note that 'http' module and 'https' module are separate modules.
  - If HTTP is okay, should I think about modifying the work from https://github.com/pulsar-edit/package-frontend/pull/49 to use 'http' module rather than 'https'? (I tested 'http' and it did work, but I went for 'https' as that's how we were using 'superagent' before my PR.)
- ~~How did you get those performance numbers from the main comment? Is there something I can run to exercise this code and see that it works about the same before/after? Then it would be easier to have insight into this PR to actually review it.~~
  - ~~(I'm not so familiar with Express, or this file, yet, so I don't know what much to say in this review other than "I know the premise is okay, because I just did it in https://github.com/pulsar-edit/package-frontend/pull/49)~~
  - UPDATE: I read the `README.md`, it's just `npm start`, of course it is 🤦 I should have read the README.
  - \[ **EDIT:** Okay, I'm getting speed numbers by running `time curl 'http://localhost:8080/?os=linux&type=linux_deb'`. And I suppose the size is size on disk, then? Probably. Okay, makes sense. \]
  - UPDATE: My results --> BEFORE: avg 551.22ms. AFTER: avg 540.88ms. Using `time curl` on localhost:8080.
  - NOTE: There is no console output on the _client/requester side_ anymore when accessing the microservice using `curl`. But there is still server-side console.log of the redirects being returned.

Anyhow, I trust the intent and the basic idea of this PR, and I assume you know this code pretty well, though I don't have much insightful commentary to add yet.

**Comment by DeeDeeG** [comment]
I actually really like the code style of the new lines of `microservices/download/index.js`. I find it less confusing than the Express stuff. 👍

Can't say I fully understand all of it, but the code looks a lot more self-explanatory, easier to guess what is going on, and frankly less going on, which is _good_ IMO.

File: microservices/download/utils.js
Comment (DeeDeeG) [inline microservices/download/utils.js:19]: Did the format of `req` change from removing `express`?

The previous `let prov = req.query.os ?? "";` was obviously fewer lines than this, but I admit the new code isn't actually hard to read. Just longer than "check if input is null, else use an empty string". So I assume there is no `req.query.os` anymore to look at.

Comment (DeeDeeG) [reply microservices/download/utils.js:19]: I don't know if `for...of` would be worth using here, as long as it works it works. And maybe the above is faster, IDK.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

`for...of` is a bit of a simpler way to iterate through an array, I tend to like `for...of`, from the perspective of an author, at least. Not an expert enough to say if there is like a performance argument to be had, or what.

File: microservices/download/utils.js
Comment (DeeDeeG) [inline microservices/download/utils.js:64]: Am I mistaken that calling `res.end()` multiple times is redundant?

I wouldn't be surprised if there is a reason for this... This is JavaScript, after all.

(Same thing for line 69 + line 70).

**Comment by confused-Techie** [comment]
@DeeDeeG To answer some of your questions here:

* Yes this does use HTTP which is not secure, but reason being is it doesn't need to be. SSL is terminated by our hosting environment before it hits the endpoint, so we don't have to worry about user facing SSL or encryption at all, it's all handled for us by the hosts reverse proxy. But we still do want to use `https` when reaching out from here as that's our own request and own responsibility, so you are good using `https` for your previous PR.
* As for the performance numbers, since I don't have access to the `time` utility on Windows and there is no good testing platform baked into this repo or microservice yet, I just testing by running the server locally and using Postman to hit the endpoint over and over, grabbing some average values. Not super scientific, but easy to test on your system if you'd like to give it a try.
* For the query parameters you are exactly right. They were an implementation from Express, with using `http` our query parameters are all within the URL itself, thats why you see me splitting up the actual path and query parameters in line 7 of `microservices/download/index.js`. So it is a bit more verbose to check our query parameters now. Which if we need to do much more of it, we could always create a dedicated function to handle this for us, so the code doesn't have to be repeated. But two uses side by side, seem simple enough to not warrant it just yet.
* But for `for...of` I will check that out. I mean really I can't imagine it's any large performance hit even if it is one. So simpler code is just a fine justification for using that instead. Since you are right, the current solution is a bit verbose.
* Lastly, you are totally correct, calling `end()` twice is pointless. The secondary `.end()` would likely never be reached, so it should be removed.

But thanks a ton for the review, and for thinking the code style is looking good over all. What can I say, after seeing your PR I was inspired to have a dependency-free microservice, as in the world of JavaScript and NPM that's so infrequent it's awesome. But yeah otherwise pretty familiar with this section, as I think it's only really been touched by me with @Spiker985 having written the GraphQL queries, so from my testing everything still works as we would expect. Although might be worth it to prior to bumping the version to add some testing. (Although suppose if it's done after the fact that doesn't really matter, since the testing wouldn't be in the final microservice anyway).

Again thanks for the review, I'll check out some of those simplification ideas you recommended, and see maybe if there's any more solid way to interact with query parameters to keep things less beefy there if possible. (Of course there's a million NPM modules to do this, but defeats my whole idea here. So ideally I can find some MIT licensed work within a good one and look at just a function implementation)

**Review by DeeDeeG — APPROVED** [review]
Yeah, no worries on the minor suggestions I had. They're not things that would make it run worse.

I updated my initial comment above with stuff as I figured it out, just now seeing your reply.

Approving because: I was able to test it, it works. Addressing review comments is optional IMO since it's working as-is and the suggestions were so small, I leave that ball in your court, but it's good enough for me already. If there's nothing new, this approve still applies after whatever changes that specifically/narrowly (and reasonably, lol) address the review comments. But I'll gladly take another look at any changes if you want.

Neat stuff, I hope this saves us some CPU/memory charges? Maybe? Or at least we'll just know in the back of our heads that the code is nicer now, lol. (And it was a learning experience for me, anyhow.)

File: microservices/download/utils.js
Comment (DeeDeeG) [reply microservices/download/utils.js:64]: [COMMENTARY, no need to read or address this, optional to view]: If I had to guess, you initially wrote it as `res.write()` but refactored the post content into `res.end()`? I know I looked into that myself for pr #49. I wasn't able to get it to work, for some reason.

File: microservices/download/utils.js
Comment (DeeDeeG) [reply microservices/download/utils.js:19]: Best I can do to simplify + subjectively more intuitive variable names is:

```js
function query_os(query_string) {
  // query_string should be an "&"-separated list of any number of query params

  let parsed_os = undefined;

  const params = query_string.split("&")

  for (const param of params) {
    if (param.startsWith("os=")) {
      parsed_os = param.split("=")[1];
      break;
    }
  }

  let valid = [ "linux", "arm_linux", "silicon_mac", "intel_mac", "windows" ];

  return valid.includes(parsed_os) ? parsed_os : false;
}
```

_(I know the JS convention is camelCase, but I saw underscores and went with it.)_

Dropping this in place works (on my machine, anyway).

If you like it, you can have it and/or I will push it to the branch for you if requested, but the existing version in the PR is good as-is anyhow, no worries.

ALSO: I volunteer to go ahead and fix the merge conflict real quick if you want, just let me know and I'll do it.

(Felt a bit bad holding this up with review, so whatever I can do to get it across the finish line. 👍 )

**Comment by confused-Techie** [comment]
Alright @DeeDeeG I've gone ahead and resolved the pointers you had (I know not needed but would rather get to it now)

And otherwise will go ahead and merge this one.

Thanks for your review, and approval! This microservice is now dependency free!

**Pull request merged**
